src/persist-txn/src/txn_cache.rs

```diff
@@ -579,97 +529,106 @@ impl<T: Timestamp + Lattice + TotalOrder + StepForward + Codec64> TxnsCacheState
                 .batch_idx
                 .remove(&batch)
                 .expect("invariant violation: batch should exist");
-            let prev = self
+            let (prev_data_id, prev_batch, prev_ts) = self
                 .unapplied_batches
                 .remove(&idx)
                 .expect("invariant violation: batch index should exist");
-            debug_assert_eq!(data_id, prev.0);
-            debug_assert_eq!(batch, prev.1);
+            debug_assert_eq!(data_id, prev_data_id);
+            debug_assert_eq!(batch, prev_batch);
             // Insertion timestamp should be less equal retraction timestamp.
-            debug_assert!(prev.2 <= ts);
+            debug_assert!(prev_ts <= ts);
         } else {
             unreachable!("only +1/-1 diffs are used");
         }
+        self.compact_data_times(&data_id);
+        debug_assert_eq!(self.validate(), Ok(()));
     }
 
-    fn compact_data_times(&mut self) {
```

Comment:
I think we could basically keep the old impl of this, but replacing since_ts with min_unapplied_ts? (Or something like that)

Reply:
Done

Reply:
I had to revert to my original implementation of compaction. Now that we must keep at least one write per shard, no matter how old, the old compaction method does not work. The min element in the `BinaryHeap` will be held back by the oldest write in the cache, which can be arbitrarily old. That will hold back compaction of all shards to that time.